File: services/hr-core/src/services/authService.ts
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import type { AppConfig } from "../config/env.js";
import { AppError } from "../domain/errors.js";
import { ACCESS_COOKIE_NAME, type HRUser } from "../domain/types.js";
import type { InMemoryRepository } from "../infra/repository.js";

type JwtPayload = {
  sub: string;
};

export class AuthService {
  constructor(
    private readonly repo: InMemoryRepository,
    private readonly config: AppConfig
  ) {}

  async register(input: { email: string; login: string; password: string; invite_code: string }): Promise<HRUser> {
    if (input.invite_code !== this.config.inviteCode) {
      throw new AppError(403, "Invalid invitation code");
    }

    if (this.repo.findUserByEmail(input.email) || this.repo.findUserByLogin(input.login)) {
      throw new AppError(409, "User with this email or login already exists");
    }

    const passwordHash = await bcrypt.hash(input.password, 12);
    return this.repo.createUser({
      email: input.email,
      login: input.login,
      passwordHash
    });
  }

  async authenticate(input: { email: string; password: string }): Promise<HRUser> {
    const user = this.repo.findUserByEmail(input.email);
    if (!user || !(await bcrypt.compare(input.password, user.passwordHash))) {
      throw new AppError(401, "Invalid email or password");
    }
    return user;
  }

  createAccessToken(user: HRUser): string {
    return jwt.sign({ sub: user.id } satisfies JwtPayload, this.config.jwtSecret, {
      expiresIn: this.config.jwtExpiresInSeconds
    });
  }

  getUserFromToken(token: string | undefined): HRUser {
    if (!token) {
      throw new AppError(401, "Not authenticated");
    }

    try {
      const payload = jwt.verify(token, this.config.jwtSecret) as JwtPayload;
      const user = this.repo.findUserById(payload.sub);
      if (!user) {
        throw new AppError(401, "User not found");
      }
      return user;
    } catch (error) {
      if (error instanceof AppError) throw error;
      throw new AppError(401, "Invalid token");
    }
  }

  cookieOptions() {
    return {
      httpOnly: true,
      secure: this.config.cookieSecure,
      sameSite: "lax" as const,
      path: "/",
      maxAge: this.config.jwtExpiresInSeconds
    };
  }

  cookieName(): typeof ACCESS_COOKIE_NAME {
    return ACCESS_COOKIE_NAME;
  }
}

File: services/hr-core/app/services/auth_service.py
from fastapi import HTTPException, status
from sqlalchemy import or_, select
from sqlalchemy.orm import Session

from app.core.config import HR_INVITE_CODE
from app.core.security import hash_password, verify_password
from app.models.hr_user import HRUser
from app.schemas.auth import HRLoginRequest, HRRegisterRequest


def register_hr(db: Session, payload: HRRegisterRequest) -> HRUser:
    if payload.invite_code != HR_INVITE_CODE:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Invalid invitation code")

    existing = db.scalar(
        select(HRUser).where(or_(HRUser.email == str(payload.email).lower(), HRUser.login == payload.login))
    )
    if existing is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="User with this email or login already exists")

    user = HRUser(
        email=str(payload.email).lower(),
        login=payload.login,
        password_hash=hash_password(payload.password),
    )
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def authenticate_hr(db: Session, payload: HRLoginRequest) -> HRUser:
    email = str(payload.email).lower()
    user = db.scalar(select(HRUser).where(HRUser.email == email))
    if user is None or not verify_password(payload.password, user.password_hash):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password")
    return user
